### PR TITLE
Aduser_files branch WIP

### PR DESCRIPTION
**Pull request by apalski**
This is my breakable toy project for Academy. I have started with the user files and related files.

This branch creates the user model specs, user model, user controller and sessions controller. The user model also creates a user using omniauth with Facebook, when used, and has specs for this in spec/features/user_omniauth.rb. To enable testing of omniauth some view files have been created: application/oauthError.html.erb(temporary), application/welcome.html.erb, layouts/_navigation.html.erb and sessions/new.html.erb.

The session controller and user controller are a next-up refactor project so please leave that to me at this point.

I would mainly like feedback on my two spec files, spec/features/user_omniauth.rb and spec/models/user_spec.rb. Pointers and hints would be great.
@MelissaKaulfuss, @scara82, @jonathanly, @elle , @mchapman17

### CODE REVIEW

File: app/helpers/session_helper.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    if auth_hash = request.env['omniauth.auth']
+      @user = User.create_with_omniauth(auth_hash)
+      session_id
+    else
+      not_omniauth
+    end  
+  end
+
+  def not_omniauth
+    @user = User.find_by(:name => params[:user][:name])
```
Comment (elle) [inline app/helpers/session_helper.rb]: Which hash notation do you want to use in the app?

File: app/helpers/session_helper.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    if @user && @user.authenticate(params[:user][:password])
+      session_id
+    else
+      flash[:notice] = "You couldn't be signed in, please check your name and password and try again"
```
Comment (elle) [inline app/helpers/session_helper.rb]: Flash messages should use translations

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+	validates :name, :email, presence: true
+	validates :name,:email, uniqueness: true
+
+	def self.create_with_omniauth(auth_hash)
+    user = User.find_or_create_by(uid: auth_hash["uid"])
+    user.email = auth_hash["user_info"]["email"]
+    user.password = SecureRandom.base64(10)
+    user.name = auth_hash["user_info"]["name"]
+    user.save! unless User.exists?(user.uid)
```
Comment (elle) [inline app/models/user.rb]: Use multiline `unless`

File: app/views/application/_errors.html.erb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+
+<% if target.errors.any? %>
+	<div id="error_explanation" >
+		<h2>Please correct the following <%= pluralize(target.errors.count, "error") %>:</h2>
```
Comment (elle) [inline app/views/application/_errors.html.erb]: Please use translations

File: app/views/application/oauthError.html.erb [hunk 5]
```diff
@@ -0,0 +1 @@
+<h1>Authentication failed.</h1>
```
Comment (elle) [inline app/views/application/oauthError.html.erb]: Translations

File: app/views/application/welcome.html.erb [hunk 6]
```diff
@@ -0,0 +1 @@
+<h1>Welcome to Budget App <%= @user.name %></h1>
```
Comment (elle) [inline app/views/application/welcome.html.erb]: Translations

File: app/views/layouts/_navigation.html.erb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+<nav>
+	<h3><%= link_to "Home", root_path %> |
+  <%= link_to "Sign up", new_user_path %> |
+	<%= link_to "Sign in", new_session_path %> |
+  <%= link_to "Sign out", '/sessions/destroy', :method => :delete %> |
+	<%= link_to "Sign in with Facebook", '/auth/facebook' %></h3>
```
Comment (elle) [inline app/views/layouts/_navigation.html.erb]: All link keys should be using translations.

File: app/views/layouts/_navigation.html.erb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+<nav>
+	<h3><%= link_to "Home", root_path %> |
+  <%= link_to "Sign up", new_user_path %> |
+	<%= link_to "Sign in", new_session_path %> |
+  <%= link_to "Sign out", '/sessions/destroy', :method => :delete %> |
```
Comment (elle) [inline app/views/layouts/_navigation.html.erb]: Double quotes

File: app/views/layouts/_navigation.html.erb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+<nav>
+	<h3><%= link_to "Home", root_path %> |
```
Comment (elle) [inline app/views/layouts/_navigation.html.erb]: Is it semantically correct to use a `h3` for a navlist?

File: app/views/layouts/_navigation.html.erb (hunk 8 above)
Comment (elle) [inline app/views/layouts/_navigation.html.erb]: `Sign out` link should only be shown is a use is logged in

File: app/views/layouts/application.html.erb [hunk 10]
```diff
@@ -4,17 +4,13 @@
...
-    config/locales/en.yml. For more see:
-    https://github.com/calebthompson/title#usage
-  %>
   <title><%= title %></title>
   <%= stylesheet_link_tag :application, media: "all" %>
   <%= csrf_meta_tags %>
 </head>
 <body class="<%= body_class %>">
   <%= render "flashes" -%>
+  <%= render 'layouts/navigation' %>
```
Comment (elle) [inline app/views/layouts/application.html.erb]: Double quotes

File: app/views/sessions/_form.html.erb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<div class="field" >
+	<%= f.label :name %>
+	<%= f.text_field :name %><br><br>
```
Comment (elle) [inline app/views/sessions/_form.html.erb]: Instead of using `<br><br>`, consider using styles appropriately to style the form to look how you want it to look.
Adding the break tags means adding presentation logic into your html markup.

File: app/views/sessions/_form.html.erb [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+</div>
+<div class="field" >
+	<%= f.label :password %>
+	<%= f.password_field :password %><br><br>
```
Comment (elle) [inline app/views/sessions/_form.html.erb]: Same here

File: app/views/sessions/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+<h1>Please sign in:</h1>
```
Comment (elle) [inline app/views/sessions/new.html.erb]: Please use translations

File: app/views/sessions/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+
+<%= form_for :user, url: '/sessions' do |f| %>
+  <%= render 'form', f: f %>
```
Comment (elle) [inline app/views/sessions/new.html.erb]: Double quotes

File: app/views/sessions/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,8 @@
...
+	<%= f.submit "Sign In" %>
+<% end %>
+
+<h2>Or sign in with Facebook above</h2>
```
Comment (elle) [inline app/views/sessions/new.html.erb]: Translations

File: app/views/sessions/new.html.erb [hunk 16]
```diff
@@ -0,0 +1,8 @@
+<h1>Please sign in:</h1>
+
+<%= form_for :user, url: '/sessions' do |f| %>
+  <%= render 'form', f: f %>
+	<%= f.submit "Sign In" %>
```
Comment (elle) [inline app/views/sessions/new.html.erb]: Translations

File: app/views/users/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+<h2>To become a member please complete your details below</h2>
+
+<%= form_for @user do |f| %>
+	<%= render 'application/errors', target: @user %>
+  <div class="field" >
+	<%= f.label :email %>
+	<%= f.text_field :email %><br><br>
+</div>	
+	<%= render 'sessions/form', f: f %>
+	<%= f.submit %>
```
Comment (elle) [inline app/views/users/new.html.erb]: Translations and double quotes please

File: app/views/users/new.html.erb [hunk 18]
```diff
@@ -0,0 +1,14 @@
+<h1>Welcome to BudgetApp</h1>
+
+<h2>If you are already a member please signin above</h2>
+<h2>To become a member please complete your details below</h2>
+
+<%= form_for @user do |f| %>
+	<%= render 'application/errors', target: @user %>
+  <div class="field" >
+	<%= f.label :email %>
+	<%= f.text_field :email %><br><br>
```
Comment (elle) [inline app/views/users/new.html.erb]: No need for the break tags

File: app/views/users/new.html.erb [hunk 19]
```diff
@@ -0,0 +1,14 @@
+<h1>Welcome to BudgetApp</h1>
+
+<h2>If you are already a member please signin above</h2>
+<h2>To become a member please complete your details below</h2>
```
Comment (elle) [inline app/views/users/new.html.erb]: Should these be `h2` or would you consider styling them bigger with stylesheets?

File: config/initializers/omniauth.rb [hunk 20]
```diff
@@ -0,0 +1,5 @@
+Rails.application.config.middleware.use OmniAuth::Builder do
+  provider :facebook, ENV['FACEBOOK_KEY'], ENV['FACEBOOK_SECRET']
```
Comment (elle) [inline config/initializers/omniauth.rb]: Double quotes

File: config/routes.rb [hunk 21]
```diff
@@ -1,2 +1,11 @@
 Rails.application.routes.draw do
+
+  root "application#welcome"
```
Comment (elle) [inline config/routes.rb]: Consider creating a "WelcomeController#show` instead, which will follow the RESTful convention.

Also, another convention is to place the default root path as the last declaration in the routes file.

File: config/secrets.yml [hunk 22]
```diff
@@ -1,8 +1,15 @@
-shared:
-  secret_key_base: <%= ENV["SECRET_KEY_BASE"] %>
 
 development:
+  secret_key_base: 35bba17b71c4f6ee1a4e45ab6c3917362a67b8478e45bfcaac728b7cfa717219ce0b7c3aeb67d4d98cb844c0ff29a06154b4483efa13a7cda1365fab5dad5415
+  facebook_app_id: <%= ENV["FACEBOOK_APP_ID"] %>
+  facebook_app_secret: <%= ENV["FACEBOOK_APP_SECRET"] %>
 
 test:
+  secret_key_base: 35bba17b71c4f6ee1a4e45ab6c3917362a67b8478e45bfcaac728b7cfa717219ce0b7c3aeb67d4d98cb844c0ff29a06154b4483efa13a7cda1365fab5dad5415
```
Comment (elle) [inline config/secrets.yml]: You should move these keys to `.env`.

Look into using different environments with the dotenv-rails gem.

File: db/migrate/20171015160550_create_users.rb [hunk 23]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration[5.1]
+  def change
+    create_table :users do |t|
+      t.string :name
```
Comment (elle) [inline db/migrate/20171015160550_create_users.rb]: Add `null: false` to both `name` and `email`

File: db/migrate/20171118114518_create_delayed_jobs.rb [hunk 24]
```diff
@@ -0,0 +1,22 @@
+class CreateDelayedJobs < ActiveRecord::Migration[5.1]
+  def self.up
+    create_table :delayed_jobs, force: true do |table|
```
Comment (elle) [inline db/migrate/20171118114518_create_delayed_jobs.rb]: Since you are using `t` in the other migration, I would keep it consistent here as well

File: db/migrate/20171118114518_create_delayed_jobs.rb [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+      table.integer :priority, default: 0, null: false # Allows some jobs to jump to the front of the queue
+      table.integer :attempts, default: 0, null: false # Provides for retries, but still fail eventually.
+      table.text :handler,                 null: false # YAML-encoded string of the object that will do work
```
Comment (elle) [inline db/migrate/20171118114518_create_delayed_jobs.rb]: Don't align the `null: false` to the other lines. Just keep it as one space from the comma

File: spec/features/users/current_user_spec.rb [hunk 26]
```diff
@@ -0,0 +1,41 @@
+require "rails_helper"
+require "./app/models/user"
```
Comment (elle) [inline spec/features/users/current_user_spec.rb]: You don't need to require the user file manually. Since we are calling "rails_helper", it loads the whole rails environment for us.

File: spec/features/users/current_user_spec.rb [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+
+describe "current_user" do
+  scenario "can't sign in again" do
+    user = User.create(name: "John Doe",
+                       email: "user@gmail.com",
```
Comment (elle) [inline spec/features/users/current_user_spec.rb]: Align to two spaces tab on the next line. Fore example:

```rb
user = User.create(name: "John",
  email: "john@example.com",
)
```

On the other hand, use a factory here instead.

File: spec/features/users/current_user_spec.rb [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+                       password: "secret",
+                       uid: "123")
+    user.save!
+
+    visit new_session_path
+
+    fill_in "Name", with: "John Doe"
```
Comment (elle) [inline spec/features/users/current_user_spec.rb]: Use translation keys here.
Also, you can use `fill_in "name", with: user.name`

File: spec/features/users/current_user_spec.rb [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+    fill_in "Password", with: "secret"
+    click_on "Sign In"
+
+    click_on "Sign in"
+
+    expect(page).to have_content("You are already signed in")
```
Comment (elle) [inline spec/features/users/current_user_spec.rb]: Use translations

File: spec/features/users/current_user_spec.rb [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  scenario "can't access the sign up page" do
+    user = User.create(name: "John Doe",
+                       email: "user@gmail.com",
+                       password: "secret",
+                       uid: "123")
+    user.save!
```
Comment (elle) [inline spec/features/users/current_user_spec.rb]: Since you are using `create`, the `save` here is redundant.
But as I mentioned previously, use a factory instead.

File: spec/features/users/current_user_spec.rb [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+    visit new_session_path
+
+
+    fill_in "Name", with: "John Doe"
+    fill_in "Password", with: "secret"
+    click_on "Sign In"
+
+    click_on "Sign up"
+
+    expect(page).to have_content("You are already signed in")
```
Comment (elle) [inline spec/features/users/current_user_spec.rb]: Use translations

File: spec/features/users/user_omniauth_spec.rb [hunk 32]
```diff
@@ -0,0 +1,39 @@
+require "rails_helper"
+require "./app/models/user"
```
Comment (elle) [inline spec/features/users/user_omniauth_spec.rb]: No need to call the user model here directly

File: spec/features/users/user_omniauth_spec.rb [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+      'uid' => '123545',
+      'user_info' => {
+        'name' => 'mockuser',
+        'email' => 'mock@gmail.com'
+      },
+      'credentials' => {
+        'token' => 'mock_token',
+        'secret' => 'mock_secret'
+      }
+    }
```
Comment (elle) [inline spec/features/users/user_omniauth_spec.rb]: Double quotes

File: spec/features/users/user_omniauth_spec.rb [hunk 34]
```diff
@@ -0,0 +1,39 @@
...
+
+    visit new_session_path
+
+    expect(page).to have_content("Sign in with Facebook")
```
Comment (elle) [inline spec/features/users/user_omniauth_spec.rb]: Translations

File: spec/features/users/user_omniauth_spec.rb [hunk 35]
```diff
@@ -0,0 +1,39 @@
...
+
+    click_on "Sign in with Facebook"
```
Comment (elle) [inline spec/features/users/user_omniauth_spec.rb]: The expectation on line 21 is redundant since you are now clicking on the same text

File: spec/features/users/user_omniauth_spec.rb [hunk 36]
```diff
@@ -0,0 +1,39 @@
...
+
+    expect(page).to have_content("mockuser")
+  end
+
+  scenario "can handle authentication error" do
+    OmniAuth.config.mock_auth[:facebook] = :invalid_credentials
+
+    visit new_session_path
+
+    expect(page).to have_content("Sign in with Facebook")
```
Comment (elle) [inline spec/features/users/user_omniauth_spec.rb]: Same as previous spec

File: spec/features/users/user_omniauth_spec.rb [hunk 37]
```diff
@@ -0,0 +1,39 @@
...
+
+    click_on "Sign in with Facebook"
+
+    expect(page).to have_content('Authentication failed.')
```
Comment (elle) [inline spec/features/users/user_omniauth_spec.rb]: Double quotes and please use translations

File: spec/features/users/user_signin_spec.rb [hunk 38]
```diff
@@ -0,0 +1,54 @@
+require "rails_helper"
+require "./app/models/user"
```
Comment (elle) [inline spec/features/users/user_signin_spec.rb]: No need to call the model directly

File: spec/features/users/user_signin_spec.rb [hunk 39]
```diff
@@ -0,0 +1,54 @@
...
+
+describe User do
+  scenario "can sign in to their account with correct details" do
+    user = User.create(name: "John Doe",
+                       email: "user@gmail.com",
+                       password: "secret",
+                       uid: "123")
+    user.save!
```
Comment (elle) [inline spec/features/users/user_signin_spec.rb]: Please use a factory. and also, the `save!` here is redundant

File: spec/models/user_spec.rb [hunk 40]
```diff
@@ -0,0 +1,72 @@
+require "rails_helper"
+require "./app/models/user"
+
+describe User do
+
```
Comment (elle) [inline spec/models/user_spec.rb]: No need for extra line

File: spec/models/user_spec.rb [hunk 41]
```diff
@@ -0,0 +1,72 @@
...
+  context "when all attributes are correct" do
+
```
Comment (elle) [inline spec/models/user_spec.rb]: No need for extra line

File: spec/models/user_spec.rb [hunk 42]
```diff
@@ -0,0 +1,72 @@
...
+    it "is created" do
+      user = User.create(name: "John Doe", email: "user@gmail.com", password: "secret", uid: "123")
```
Comment (elle) [inline spec/models/user_spec.rb]: Consider using `shoulda-matchers` instead to check for validations and invalid records

File: spec/models/user_spec.rb [hunk 43]
```diff
@@ -0,0 +1,72 @@
...
+
+      expect(user).to be_a User
```
Comment (elle) [inline spec/models/user_spec.rb]: If you decide to keep this spec, I would change it to be:

```rb
expect { create(:user) }.to change(User, :count).by(1)
```
```

File: spec/models/user_spec.rb [hunk 44]
```diff
@@ -0,0 +1,72 @@
...
+    end
+  end
+
+  context "when password not given" do
+
```
Comment (elle) [inline spec/models/user_spec.rb]: Extra line not needed

File: spec/models/user_spec.rb [hunk 45]
```diff
@@ -0,0 +1,72 @@
...
+    it "raises an error" do
+      user = User.create(name: "John Doe", email: "user@gmail.com", uid: "123")
```
Comment (elle) [inline spec/models/user_spec.rb]: Use a factory

File: spec/models/user_spec.rb [hunk 46]
```diff
@@ -0,0 +1,72 @@
...
+
+      expect { user.save! }.to raise_error("Validation failed: Password can't be blank")
```
Comment (elle) [inline spec/models/user_spec.rb]: Use translations

File: spec/models/user_spec.rb [hunk 47]
```diff
@@ -0,0 +1,72 @@
...
+    end
+  end
+
+  context "when name not given" do
+
```
Comment (elle) [inline spec/models/user_spec.rb]: Extra line

File: spec/models/user_spec.rb [hunk 48]
```diff
@@ -0,0 +1,72 @@
...
+    it "raises an error" do
+      user = User.create(email: "user@gmail.com", password: "secret", uid: "123")
```
Comment (elle) [inline spec/models/user_spec.rb]: You don't need to create an object in the database, which is expensive, in order to check for validation and error messages.

For example:

```rb
expect { User.new }.not_to be_valid
```

or

```rb
user = build(:user, name: nil)

user.valid?

expect(user.errors.full_messages).to include(...)
```

Ref: https://apidock.com/rails/ActiveModel/Errors/full_messages

```

File: spec/models/user_spec.rb [hunk 49]
```diff
@@ -0,0 +1,72 @@
...
+  context "when email not given" do
+
+    it "raises an error" do
+      user = User.create(name: "John Doe", password: "secret", uid: "123")
+
+      expect { user.save! }.to raise_error("Validation failed: Email can't be blank")
+    end
+  end
+
+  context "when name is not unique" do
```
Comment (elle) [inline spec/models/user_spec.rb]: You can check for uniqueness with shoulda-matchers

File: spec/models/user_spec.rb [hunk 50]
```diff
@@ -0,0 +1,72 @@
...
+    it "raises an error" do
+      user = User.create(name: "John Doe", email: "user@gmail.com", password: "secret", uid: "123")
+      user.save!
+      user2 = User.create(name: "Mary Doe", email: "user@gmail.com", password: "secret_sausage", uid: "124")
+
+      expect { user2.save! }.to raise_error("Validation failed: Email has already been taken")
+    end
+  end
+
+  context "has many budgets" do
```
Comment (elle) [inline spec/models/user_spec.rb]: I would change the context to read `associations`

File: spec/models/user_spec.rb [hunk 51]
```diff
@@ -0,0 +1,72 @@
...
+
+    xit "the association is correct" do
+      RSpec.describe User, type: :model do
+        it { should have_many(:budgets).through(:user_budgets) }
+      end
+    end
```
Comment (elle) [inline spec/models/user_spec.rb]: Instead of lines 66-70, I would have:

```rb
  it { is_expected.to have_many(:budgets).through(:user_budgets) }
```

File: spec/rails_helper.rb [hunk 52]
```diff
@@ -17,6 +18,17 @@ module Features
...
+  config.before(:suite) do
+    DatabaseCleaner.strategy = :transaction
+    DatabaseCleaner.clean_with(:truncation)
+  end
+
+  config.around(:each) do |example|
+    DatabaseCleaner.cleaning do
+      example.run
+    end
+  end
```
Comment (elle) [inline spec/rails_helper.rb]: DatabaseCleaner configurations should live in `spec/support/database_cleaner.rb`

File: spec/spec_helper.rb [hunk 53]
```diff
@@ -25,6 +24,10 @@
 
   config.example_status_persistence_file_path = "tmp/rspec_examples.txt"
   config.order = :random
+
+  config.include Capybara::DSL
```
Comment (elle) [inline spec/spec_helper.rb]: This should not live here. Why did you need to add this line?

File: spec/spec_helper.rb [hunk 54]
```diff
@@ -25,6 +24,10 @@
...
+
+  FactoryBot.allow_class_lookup = false
```
Comment (elle) [inline spec/spec_helper.rb]: Why is this line needed?
And if it is needed, it should live in `spec/support/factory_bot.rb`

File: spec/spec_helper.rb [hunk 55]
```diff
@@ -25,6 +24,10 @@
...
 end
 
 WebMock.disable_net_connect!(allow_localhost: true)
```
Comment (elle) [inline spec/spec_helper.rb]: This line should live in `spec/support/webmock.rb`

File: Gemfile [hunk 56]
```diff
@@ -39,14 +41,14 @@ group :development, :test do
   gem "bullet"
   gem "bundler-audit", ">= 0.5.0", require: false
   gem "dotenv-rails"
-  gem "factory_girl_rails"
+  gem "factory_bot_rails"
   gem "pry-byebug"
   gem "pry-rails"
```
Comment (elle) [inline Gemfile]: Since you are using "pry-byebug", this gem is redundant

File: app/controllers/application_controller.rb [hunk 57]
```diff
@@ -1,3 +1,10 @@
 class ApplicationController < ActionController::Base
+
   protect_from_forgery with: :exception
+
+  include ApplicationHelper, SessionHelper
```
Comment (elle) [inline app/controllers/application_controller.rb]: All Helpers are automatically included in recent versions of Rails

File: app/controllers/users_controller.rb [hunk 58]
```diff
@@ -0,0 +1,58 @@
+class UsersController < ApplicationController
+
+  def index
+		@users = User.all.sort_by {|user| user.name.downcase}
+	end
+
+	def new
+		if !current_user
```
Comment (elle) [inline app/controllers/users_controller.rb]: Either use `unless current_user` or consider switching the case branching

File: Gemfile [hunk 59]
```diff
@@ -63,4 +65,4 @@ end
 gem 'high_voltage'
 gem 'bourbon', '~> 5.0.0.beta.8'
 gem 'neat', '~> 2.1'
-gem 'refills', group: [:development, :test]
\ No newline at end of file
+gem 'refills', group: [:development, :test]
```
[the remark this answers is not in the document]
Comment (elle) [reply Gemfile]: Consider culling out the gems you are not planning on using, for example neat or refills

File: Gemfile [hunk 60]
```diff
@@ -39,14 +41,14 @@ group :development, :test do
   gem "bullet"
   gem "bundler-audit", ">= 0.5.0", require: false
   gem "dotenv-rails"
-  gem "factory_girl_rails"
+  gem "factory_bot_rails"
   gem "pry-byebug"
   gem "pry-rails"
-  gem "rspec-rails", "~> 3.6"
+  gem "rspec-rails", "~> 3.7"
+  gem "capybara-webkit"
```
[the remark this answers is not in the document]
Comment (apalski) [reply Gemfile]: Thanks Elle, sorry, you're right, way too big a PR!
I got pretty carried away, I've got to learn to pace myself and not overload my reviewers! 
Have checked the Gemfile order and corrected.
Thanks so much for all the work you've put into this for me, really appreciate it!

File: app/controllers/application_controller.rb [hunk 61]
```diff
@@ -1,3 +1,10 @@
...
+
+  def welcome
+		@user = current_user
+	end
```
[the remark this answers is not in the document]
Comment (apalski) [reply app/controllers/application_controller.rb]: Thanks Elle, Atom was set to soft tabs, so I've reset to hard.
Will look at how else to handle the welcome action

**Comment by apalski** [comment]
Closing PR to redo the work in new branches to enable smaller pull requests.
